File: python/binary_search/median_of_two_sorted.py
import math
from typing import List

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        if len(nums1) > len(nums2):
            return self.findMedianSortedArrays(nums2, nums1)

        x, y = len(nums1), len(nums2)

        low, high = 0, x

        while low <= high:
            partitionX = low + high // 2

            partitionY = (x + y + 1) // 2 - partitionX

            max_left_x = -math.inf

            if partitionX != 0:
                max_left_x = nums1[partitionX-1]


            min_right_x = math.inf

            if partitionX != x:
                min_right_x = nums1[partitionX]


            max_left_y = math.inf

            if partitionY != y:
                max_left_y = nums2[partitionY-1]


            min_right_y = math.inf

            if partitionY != y:
                min_right_y = nums2[partitionY]


            if max_left_x <= min_right_y and max_left_y <= min_right_x:
                if( x + y) % 2 == 0:
                    return float(max(max_left_x, max_left_y)+min(min_right_x, min_right_y)) // 2.0
                else:
                    return max(max_left_x, max_left_y)

            elif max_left_x > min_right_y:
                high = partitionX - 1
            else:
                low = partitionX + 1

        return 0.0

File: golang/binary_search/median_of_two_sorted.go
package main

import "math"

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	if len(nums1) > len(nums2) {
		return findMedianSortedArrays(nums2, nums1)
	}

	x, y := len(nums1), len(nums2)

	low, high := 0, x

	for low <= high {
		partitionX := (low + high) / 2
		partitionY := (x+y+1)/2 - partitionX

		maxLeftX := math.MinInt64
		if partitionX != 0 {
			maxLeftX = nums1[partitionX-1]
		}

		minRightX := math.MaxInt64
		if partitionX != x {
			minRightX = nums1[partitionX]
		}

		// Tetapkan nilai untuk maxLeftY dan minRightY dengan batasan partisi
		maxLeftY := math.MinInt64
		if partitionY != 0 {
			maxLeftY = nums2[partitionY-1]
		}

		minRightY := math.MaxInt64
		if partitionY != y {
			minRightY = nums2[partitionY]
		}

		// Cek kondisi keseimbangan partisi
		if maxLeftX <= minRightY && maxLeftY <= minRightX {
			// Jika total elemen genap
			if (x+y)%2 == 0 {
				return float64(max(maxLeftX, maxLeftY)+min(minRightX, minRightY)) / 2.0
			} else { // Jika total elemen ganjil
				return float64(max(maxLeftX, maxLeftY))
			}
		} else if maxLeftX > minRightY {
			high = partitionX - 1
		} else {
			low = partitionX + 1
		}
	}

	return 0.0
}

func ifElse(condition bool, trueVal, falseVal interface{}) interface{} {
	if condition {
		return trueVal
	}
	return falseVal
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
